proposals/per-extension-language-preferences.md

```diff
@@ -0,0 +1,235 @@
+  - The browser should update related extension UI, e.g. `action.default_title`.
+  - The browser might update related browser UI, but this part is up to the browser to decide.
+- Text from `i18n.getMessage()` and css files don't update. These are handled by developers.
+
+##### Persistence of `i18n.setCurrentLanguage(code)`
+
+This setting is persistent when:
+- restart or upgrade the browser.
+- disable then re-enable the extension.
+- upgrade the extension to a new version.
+
+#### Behavior of `i18n.getAllLanguages()`
+
+This method returns an array of language tags that the extension supported, e.g. `["en", "fr", "ja", "zh-CN"]`.
+This is a convenient method. Without this method, developers need to hardcode it into their code.
+
+#### Behavior of `i18n.onLanguageChanged` event
+
+When the user preferred language is changed to a new different language by `i18n.setCurrentLanguage(code)`, this event is fired.
+Developers use this event to know the language changed, and might do the following:
+
+- Update the content of extension pages in place (without refreshing the page)
+- Reload extension pages:
+  - If the page is stateless, refresh the page directly.
+  - If the page is stateful, save the current state and then refresh the page.
+  - Prompts the user that the language setting has changed, and asks the user if they want to refresh the page immediately.
+- Re-create the extension context menus, since the title of the context menus shoule be updated.
+- Update the action title and badge text if needed.
+
+#### Behavior of `i18n.getMessage()`, manifest and CSS files
+
+When this proposal is implemented, the behavior of these existing functions will change as follows:
+
+- If the preferred language is not set by `i18n.setCurrentLanguage()`, it is consistent with the existing behavior.
+- If the preferred language is set by `i18n.setCurrentLanguage()`, and the extension supports this language, then prioritize using that language.
+- If the preferred language is set by `i18n.setCurrentLanguage()`, but the extension doesn't support this language, then treat as if the preferred language is not set. This is an edge case, for example, the language was removed when the extension was upgraded.
+
+### Integrate other i18n implementations (frameworks) with browser.i18n
+
+This proposal can also be integrated with third-party i18n frameworks, as these frameworks typically allow developers to specify the language to be used.
+For example, developers use `i18n.getCurrentLanguage()` and `i18n.setCurrentLanguage()` for the user preferred language, but don't use `i18n.getMessage()`.
+
+```js
+otherI18nFramework.setLanguage(browser.i18n.getCurrentLanguage());
+
+let text = otherI18nFramework.getMessage(key);
+```
+
+### New Permissions
+
+N/A
+
+### Manifest File Changes
+
+N/A
+
+## Security and Privacy
+
+### Exposed Sensitive Data
+
+N/A
+
+### Abuse Mitigations
+
+N/A
+
+### Additional Security Considerations
```

Comment:
This proposal has `N/A` at "Abuse Mitigations" and "Additional Security Considerations".

I see potential for abuse: An extension could rapidly change their name at runtime to make it more difficult for users to discover what extension they are interacting with. The proposal does not describe any mitigations for this abuse concern.

Reply:
@Rob--W In general I believe it would be best to fetch the translations for the manifest file in the language of the containing UI. So if it will be used in the browser extension management page, it would match the language of said page. Same for the context menu items on the action button. This would prevent the issue you are describing.

Reply:
Although I don't think this poses a real security problem. However, the browser can limit the frequency of `setCurrentLanguage` method calls, which is up to the browser.

Reply:
I added  a supplementary note in the Abuse Mitigations section. let me know if it's Ok.